Introduce `-Xsource-features`, for customizing the behavior of `-Xsource:3` and `-Xsource:3-cross`

som-snytt opened the pull request:
This feature is only recommended for projects that cross-build between Scala 2.13 and Scala 3. For migrating to Scala 3, use plain `-Xsource:3`. See also https://docs.scala-lang.org/scala3/guides/migration/tooling-scala2-xsource3.html.

This PR makes Scala 3 language features opt-in under an option "buffet" or "à la carte". To view the menu:

```
$> scalac -Xsource-features:help
Enable Scala 3 features under -Xsource:3.

Instead of `-Xsource-features:_`, it is recommended to enable specific features, for
example `-Xsource-features:v2.13.14,-case-companion-function` (-x to exclude x).
This way, new semantic changes in future Scala versions are not silently adopted;
new features can be enabled after auditing the corresponding migration warnings.

`-Xsource:3-cross` is a shorthand for `-Xsource:3 -Xsource-features:_`.

Features marked with [bin] affect the binary encoding. Enabling them in a project
with existing releases for Scala 2.13 can break binary compatibility.

Available features:

  case-apply-copy-access    Constructor modifiers are used for apply / copy methods of case classes. [bin]
  case-companion-function   Synthetic case companion objects no longer extend FunctionN. [bin]
  infer-override            Inferred type of member uses type of overridden member. [bin]
  any2stringadd             Implicit `any2stringadd` is never inferred.
  unicode-escapes-raw       Don't process unicode escapes in triple quoted strings and raw interpolations.
  string-context-scope      String interpolations always desugar to scala.StringContext.
  leading-infix             Leading infix operators continue the previous line.
  package-prefix-implicits  The package prefix p is no longer part of the implicit search scope for type p.A.
  implicit-resolution       Use Scala-3-style downwards comparisons for implicit search and overloading resolution (see github.com/scala/scala/pull/6037).
  v2.13.13                  case-apply-copy-access,case-companion-function,infer-override,any2stringadd,unicode-escapes-raw,string-context-scope,leading-infix,package-prefix-implicits.
  v2.13.14                  v2.13.13 plus implicit-resolution
```

`-Xsource:3 -Xsource-features:_` (or `-Xsource:3-cross` for convenience to start a REPL) offers all "courses" as "prix fixe". It's not recommended in a build script, use `v2.13.14` instead to get migration warnings as new features are adopted.

`-Xmigration` is no longer "overloaded" to mean `-Wconf:cat=scala3-migration:w`. That incantation must be written in full.

Fixes scala/bug#12961

som-snytt commented:
To be squashed after review. Nominally (whatever that weaselly word means), I checked usages of `isScala3...` for reasonableness, but it's surprisingly difficult to verify. A clearer API would be more like `hardMigration`: either `migrationWarning` if `isScala3`, or discriminating `progressiveWarning(at2 = old_behavior, at3 = warn, at3X = cross)` where cross warns if verbose.

The `-Ysource-options:case-access` warning for `apply` is an unduly heavyweight mechanism for a niche feature or warning.

som-snytt commented:
Happy to have you progress this experiment to your liking.

Here, you find out "what's new" by `-Ysource-options:verbose` (if you triggered it) or `help` (IIRC there was help text).

I think just having `-Xnew-stuff:case-access` etc is fine, in the xlint mode; I just wanted to see if we could avoid it, because it results in "fine-tuning". The experiment here was `3-compat` allows for opt-in to dangerous options. So there was the conservative flag, the all-in flag, and the goldilocks flag.

Here, `-Xsource:3-migration` is just an alias for `-Xsource:3`, for clarity.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1128,7 +1128,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
 
       def doAdaptApplyInsertion(retry: Boolean): Tree =
-        if (currentRun.isScala3Cross && !isPastTyper && tree.symbol != null && tree.symbol.isModule && tree.symbol.companion.isCase && isFunctionType(pt))
+        if (currentRun.isScala3X && !isPastTyper && tree.symbol != null && tree.symbol.isModule && tree.symbol.companion.isCase && isFunctionType(pt))
```
I realize that apply insertion should be enabled by default, not even under `-Xsource:3`.

For a library built with `-Xsoruce:3-cross`, case class companions don't extend `FunctionN`. This should not break `List(1).map(A)` for users of the library that use the default compiler options.

@som-snytt wdyt?

som-snytt replied:
The library would use `3-compat` (or equivalent) to avoid incompatible changes.

The librarian, for the same reason, would not add `object A` companion that has the same effect; or would be sorry if they did.

But you may be right that this should be considered a "benign" feature change that everyone could benefit from. Fork all the language features! It is of general utility. The downside is your 2.12 build breaks; but who does that? I guess only middle-size fish who consume smaller fish and are consumed in turn.

lrytz commented:
We might rename `-Ysource-options`

Also I could probably live with an alias `Xsource:3-cross` for `-Ysource-options:_`, it would be handy for testing stuff (not just for us, also for actual users of `Ysource-options` when they want to fire up a REPL).

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Unapplies.scala:
```diff
@@ -98,25 +98,31 @@ trait Unapplies extends ast.TreeDSL {
-    else if (currentRun.isScala3 && changeModsIn3) Warn
+    if (!changeModsIn3) Default
+    else if (currentRun.scala3Features.caseApplyCopyAccess) Inherit
+    else if (currentRun.isScala3) Warn
```
These changes were to say "both inherit and warn about it". That's not needed because why? If they ask for it via option, it should never warn. Oder?

lrytz replied:
That's how I see it, there should no warning for using features that you opted in.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3477,7 +3484,9 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
                 newStats += typedStat(tree) // might add even more synthetics to the scope
                 tree.getAndRemoveAttachment[CaseApplyInheritAccess.type].foreach(_ =>
-                  runReporting.warning(tree.pos, "access modifiers for `apply` method are copied from the case class constructor", Scala3Migration, sym))
+                  runReporting.warning(tree.pos, "access modifiers for `apply` method were copied from the case class constructor under Scala 3", Scala3Migration, sym))
```
Wasn't sure the extra attachment would be needed if you only get cross-features you asked for, but now I'm not sure what the warning policy is. I will continue reading...

lrytz replied:
One of the two annotations is stale now, it's either warn or new semantics

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Unapplies.scala:
```diff
@@ -163,10 +169,13 @@ trait Unapplies extends ast.TreeDSL {
-      if (access == Warn) m.updateAttachment(CaseApplyInheritAccess))
+      if (ApplyAccess.isWarn(access))
+        if (ApplyAccess.isInherit(access)) m.updateAttachment(CaseApplyInheritAccess)
+        else m.updateAttachment(CaseApplyInheritAccessMigration)
```
IIUC it is no longer both warn + inherit, so only one attachment is needed?

som-snytt commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -1180,12 +1180,23 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
+      def leadingInfix           = isScala3 && contains(o.leadingInfix)
+      def packagePrefixImplicits = isScala3 && contains(o.packagePrefixImplicits)
+      def implicitResolution     = isScala3 && contains(o.implicitResolution) || settings.Yscala3ImplicitResolution.value
+    }
```
A "choice" use of the facility!  Also nice to fold that "implicit resolution" option into the menu.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Unapplies.scala:
```diff
@@ -285,11 +294,10 @@ trait Unapplies extends ast.TreeDSL {
           else {
             if (mods3 != synth)
-              runReporting.warning(cdef.pos, "access modifiers for `copy` method are copied from the case class constructor", Scala3Migration, cdef.symbol)
+              runReporting.warning(cdef.namePos, "access modifiers for `copy` method will be copied from the case class constructor under Scala 3", Scala3Migration, cdef.symbol)
```
this was a tweak to tighten the span of the diagnostic. Eventually someone using metals would notice.

The apply method is still just the tree focus because I was too lazy to downcast. Or I was too downcast.

lrytz committed:
Make any2stringadd a Scala 3 migraiton warning

Co-authored-by: A. P. Marki <som.snytt@gmail.com>

lrytz committed:
Support .apply insertion for case companions without Xsource:3

A dependency might be built with
-Xsource-features:case-companion-function.
Apply insertion should work for users that don't enable that flag.

Co-authored-by: A. P. Marki <som.snytt@gmail.com>

lrytz committed:
"Synthetic case companion used as a Function" warning for selections

Move the warning to adapt. This makes it work also for selections,
not just idents.

Co-authored-by: A. P. Marki <som.snytt@gmail.com>

lrytz committed:
Fix spurious warning for inferred method change under -Xsource:3

Workaround for scala/bug 12968.

Co-authored-by: A. P. Marki <som.snytt@gmail.com>

lrytz committed:
Add -Xsource-features to replace -Xsource:3-cross

This allows enabling Scala 3 semantics for cross-building
individually.

Co-authored-by: A. P. Marki <som.snytt@gmail.com>

lrytz commented:
I squashed almost everything into one commit, only a few semantic changes as separate commits as good as possible.

som-snytt commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -144,31 +147,72 @@ trait ScalaSettings extends StandardScalaSettings with Warnings { _: MutableSett
+          |`-Xsource:3-cross` is a shorthand for `-Xsource:3 -Xsource-features:_`.
+          |
+          |Available features:
+          |""".stripMargin)
```
extra strip (it would be interesting if that failed to compile)

lrytz replied:
👍 IntelliJ adds it when typing `sm"""<return>`...

som-snytt commented on test/files/run/print-args.scala:
```diff
@@ -26,6 +27,7 @@ object Test extends DirectTest {
     """
   def show() = {
     assert(!compile())
+    //assertEquals(expected.linesIterator.toList, Files.readAllLines(argfile).asScala.toList)
```
debug cruft. The `expected` needs `.trim` and assert could be
```
    import scala.tools.testkit.AssertUtil.assertSameElements
    assertSameElements(expected.linesIterator, Files.readAllLines(argfile).asScala)
```
The test is because `-Vprint-args` was broken; the `-Xsource` is not relevant.

I'll submit the tweak separately. :)

som-snytt commented:
The buffet option makes it easier to see in the source where the features take effect.

It also simplifies the warnings under the rule of "no warning if they asked for a feature". (I just remembered that the point of `verbose` mode was to enable "audit", so you could see "this override was inferred differently", for example. Reverting to migration might show the future warning, but it might also just show an error?) (Probably no one would ever use verbose audit anyway.)

It took me two tries to correctly test moving the "case function" check from ident to adapt. Very nice!

som-snytt commented:
LGTM. Lukas pulled that old github mind trick where he gets me to open the PR so that only he can approve it later. :)

som-snytt commented:
travis claims
```
-t12919.scala:48: error: Implicit method myClassToSeq was found in a package prefix of the required type, which is not part of the implicit scope in Scala 3.

+t12919.scala:48: error: Implicit method myClassToSeq was found in a package prefix of the required type, which is not part of the implicit scope in Scala 3 (or with -Xsource-features:package-prefix-implicits).
```

lrytz commented:
semantic merge conflict with https://github.com/scala/scala/pull/10706, rebased.

> Your old trick of emitting the "how to suppress it" message would be helpful

Maybe somebody will eventually figure out how to do interactive verbose (not "re-run with ...").

```
➜ sandbox s -Wconf:any:wv
> scala repl -S 2.13 -cp . -Wconf:any:wv
Welcome to Scala 2.13.13 (OpenJDK 64-Bit Server VM, Java 21.0.1).
Type in expressions for evaluation. Or try :help.

scala> implicit val x = 41
                    ^
       warning: Implicit definition should have explicit type (inferred Int) [quickfixable]
       Applicable -Wconf / @nowarn filters for this warning: msg=<part of the message>, cat=other-implicit-type, site=x
val x: Int = 41
```

SethTisue commented:
I'm noticing in the community build that despite this change:

> -Xmigration is no longer "overloaded" to mean -Wconf:cat=scala3-migration:w. That incantation must be written in full.

I'm still seeing errors like:

```
[playframework] [error] Scala 3 migration messages are errors under -Xsource:3. Use -Wconf / @nowarn to filter them or add -Xmigration to demote them to warnings.
```

seems like an oversight that this advice wasn't updated?

som-snytt commented:
I never read those messages all the way to the end.

SethTisue commented:
a bit of good news: over in the community build, I removed `-Xmigration` from about a two dozen repos and only had to replace it with `-Wconf:cat=scala3-migration:w` in five:

```
proj/blaze.conf:    "appendScalacOptions -Wconf:cat=scala3-migration:w"
proj/json4s.conf:    "appendScalacOptions -Wconf:cat=scala3-migration:w"
proj/mima.conf:    "appendScalacOptions -Wconf:cat=scala3-migration:w"
proj/playframework.conf:    "appendScalacOptions -Wconf:cat=scala3-migration:w"
proj/scalafx.conf:    "appendScalacOptions -Wconf:cat=scala3-migration:w"
```

ghostbuster91 commented:
Hi, this is really great feature. Unfortunately many projects that cross build for 2.13 also cross-build for 2.12. Are there any chances/plans of getting this backported into 2.12 line?

SethTisue commented:
@ghostbuster91 in short, no
